### PR TITLE
Send Analytics Logs to DB

### PR DESCRIPTION
**Pull request by leszko**
Implements enriching the analytics log, processing it, and pushing as metrics into Victoria Metrics (actually any DB supporting Prometheus format).

Sorry for this PR to be that big 🙃 

**Changes**
- Fetch studio-related data to enrich analytics log
- Process the log and send it into the Victoria Metrics

Tech Design for the context: https://www.notion.so/livepeer/Studio-2-0-Realtime-Viewership-Backend-2511fd91b0854d09958eed1fa548f447

fix https://linear.app/livepeer/issue/ENG-1650/enrich-event-data-with-studio-data-and-publish-to-db

### CODE REVIEW

File: mapic/mistapiconnector_app.go [hunk 1]
```diff
@@ -810,6 +815,18 @@ func (mc *mac) refreshStreamInfo(playbackID string) (*streamInfo, error) {
 	return info, nil
 }
 
+func (mc *mac) GetCachedStream(playbackID string) *api.Stream {
```
Comment (victorges) [inline mapic/mistapiconnector_app.go:818]: An alternative to the refactor I proposed would be exposing the `getStreamInfo` function as this `GetCachedStream` (without `Cached` I guess), then the actual `log_ext_data.go` would only need to handle assets.

Still optional though, just thinking how to keep the "studio is just a provider" reality a lil closer.

**Review by victorges — APPROVED** [review]
LGTM

File: handlers/analytics.go [hunk 2]
```diff
@@ -41,35 +48,53 @@ type AnalyticsLogEvent struct {
...
+		extFetcher:   analytics.NewExternalDataFetcher(streamCache, lapi),
+		logProcessor: analytics.NewLogProcessor(metricsURL, host),
+	}
 }
 
-func (d *AnalyticsHandlersCollection) Log() httprouter.Handle {
+func (c *AnalyticsHandlersCollection) Log() httprouter.Handle {
 	schema := inputSchemasCompiled["AnalyticsLog"]
 
+	dataCh := make(chan analytics.LogData, MaxConcurrentProcessing)
```
[the remark this answers is not in the document]
Comment (leszko) [reply handlers/analytics.go:71]: Updated

File: handlers/analytics.go [hunk 3]
```diff
@@ -129,3 +153,35 @@ func getOrAddMissing(key string, headers http.Header, missingHeaders []string) (
...
+
+func deviceTypeOf(ua useragent.UserAgent) string {
+	if ua.Mobile {
+		return "mobile"
+	} else if ua.Tablet {
+		return "tablet"
+	} else if ua.Desktop {
+		return "desktop"
+	}
+	return ""
```
[the remark this answers is not in the document]
Comment (leszko) [reply handlers/analytics.go:186]: Updated

File: handlers/analytics/log_ext_data.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		ExternalData{
+			UserID: asset.UserID,
+		})
+}
+
+func (e *ExternalDataFetcher) cacheExtData(playbackID string, extData ExternalData) (ExternalData, error) {
+	e.mu.Lock()
+	e.cache[playbackID] = extData
+	e.mu.Unlock()
+	return extData, nil
```
[the remark this answers is not in the document]
Comment (leszko) [reply handlers/analytics/log_ext_data.go:93]: Updated

File: handlers/analytics/log_ext_data.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	e.mu.RUnlock()
+	if ok {
+		// Empty struct means that the playbackID does not exist
+		if cached == (ExternalData{}) {
+			return cached, fmt.Errorf("playbackID does not exists, playbackID=%v", playbackID)
+		}
+		return cached, nil
+	}
+
+	// PlaybackID is not in internal cache, try using the stream cache from mapic
```
[the remark this answers is not in the document]
Comment (leszko) [reply handlers/analytics/log_ext_data.go:48]: Yeah, let me keep it as it is. Not that I don't like your idea, but each has drawbacks, so probably not worth changing right now.

File: mapic/mistapiconnector_app.go (hunk 1 above)
Comment (leszko) [reply mapic/mistapiconnector_app.go:818]: Yeah, let's maybe leave thinking about this more if we decide to have "studio as a provider". This part will definitely not make it harder 🙃

File: handlers/analytics.go [hunk 6]
```diff
@@ -41,35 +48,53 @@ type AnalyticsLogEvent struct {
...
 		log, err := parseAnalyticsLog(r, schema)
 		if log == nil {
-			errors.WriteHTTPBadRequest(w, "Invalid request payload", err)
+			cerrors.WriteHTTPBadRequest(w, "Invalid request payload", err)
 			return
 		}
 		geo, err := parseAnalyticsGeo(r)
 		if err != nil {
-			glog.Warning("error parsing geo info from analytics log request header, err=%w", err)
+			glog.Warning("error parsing geo info from analytics log request header, err=%v", err)
```
Comment (pwilczynskiclearcode) [inline handlers/analytics.go:82]: @leszko Let's not use "error" word if it's just a warning. This is misleading.

**Pull request merged**